cranelift-codegen/src/isa/x86/abi.rs

```diff
@@ -791,12 +917,46 @@ fn insert_common_epilogues(
     stack_size: i64,
     reg_type: ir::types::Type,
     csrs: &RegisterSet,
+    isa: &dyn TargetIsa,
+    cfa_state: Option<CFAState>,
 ) {
     while let Some(ebb) = pos.next_ebb() {
         pos.goto_last_inst(ebb);
         if let Some(inst) = pos.current_inst() {
             if pos.func.dfg[inst].opcode().is_return() {
-                insert_common_epilogue(inst, stack_size, pos, reg_type, csrs);
+                // Figure out if we need to insert end-of-function-aware frame layout information.
+                let following_inst = pos
+                    .func
+                    .layout
+                    .next_ebb(ebb)
+                    .and_then(|next_ebb| pos.func.layout.first_inst(next_ebb));
+
+                if following_inst.is_some() {
+                    if let Some(ref mut frame_layout) = pos.func.frame_layout {
+                        frame_layout
+                            .instructions
+                            .insert(inst, vec![FrameLayoutChange::Preserve].into_boxed_slice());
+                    }
+                }
+
+                insert_common_epilogue(
+                    inst,
+                    stack_size,
+                    pos,
+                    reg_type,
+                    csrs,
+                    isa,
+                    cfa_state.clone(),
+                );
+
+                if let Some(following_inst) = following_inst {
+                    if let Some(ref mut frame_layout) = pos.func.frame_layout {
```

Comment:
The suggested change above might require to invert the two conditions here too.

Reply:
I moved this into the conditional above, since at that point we know:
* we want to record frame information
* where the epilogue starts
* where the epilogue ends (following instruction, if any)

and instruction numbers are stable indices, so even though we'll add instructions in the epilogue the end stays the same.